ompi/mca/part/persist/part_persist.h

```diff
@@ -0,0 +1,595 @@
+                         opal_cache_line_size,
+                         OBJ_CLASS(mca_part_persist_precv_request_t),
+                         0,opal_cache_line_size,
+                         ompi_part_persist.free_list_num,
+                         ompi_part_persist.free_list_max,
+                         ompi_part_persist.free_list_inc,
+                         NULL, 0, NULL, NULL, NULL);
+    opal_free_list_init (&mca_part_base_psend_requests,
+                         sizeof(mca_part_persist_psend_request_t),
+                         opal_cache_line_size,
+                         OBJ_CLASS(mca_part_persist_psend_request_t),
+                         0,opal_cache_line_size,
+                         ompi_part_persist.free_list_num,
+                         ompi_part_persist.free_list_max,
+                         ompi_part_persist.free_list_inc,
+                         NULL, 0, NULL, NULL, NULL);
+     ompi_part_persist.progress_list = OBJ_NEW(opal_list_t);
+}
+
+__opal_attribute_always_inline__ static inline void
+mca_part_persist_complete(struct mca_part_persist_request_t* request)
+{
+    if(MCA_PART_PERSIST_REQUEST_PRECV == request->req_type) {
+        request->req_ompi.req_status.MPI_SOURCE = request->req_peer; 
+    } else {
+        request->req_ompi.req_status.MPI_SOURCE = request->req_comm->c_my_rank;
+    }
+    request->req_ompi.req_complete_cb = NULL;
+    request->req_ompi.req_status.MPI_TAG = request->req_tag;  
+    request->req_ompi.req_status._ucount = request->req_bytes;
+    request->req_ompi.req_status.MPI_ERROR = OMPI_SUCCESS;
+    request->req_part_complete = true;
+    ompi_request_complete(&(request->req_ompi), true );
+}
+
+/**
+ * mca_part_persist_progress is the progress function that will be registered. It handles 
+ * both send and recv request testing and completion. It also handles freeing requests,
+ * after MPI_Free is called and the requests have become inactive.
+ */
+__opal_attribute_always_inline__ static inline int
+mca_part_persist_progress(void)
+{
+    mca_part_persist_list_t *current;
+    int err;
+    size_t i;
+
+    /* prevent re-entry, */
+    int block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), 1);
+    if(1 < block_entry)
+    {
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    OPAL_THREAD_LOCK(&ompi_part_persist.lock);
+ 
+    mca_part_persist_request_t* to_delete = NULL;
+
+    /* Don't do anything till a function in the module is called. */
+    if(-1 == ompi_part_persist.init_world)
+    {
+        OPAL_THREAD_UNLOCK(&ompi_part_persist.lock);
+        block_entry = opal_atomic_add_fetch_32(&(ompi_part_persist.block_entry), -1);
+        return OMPI_SUCCESS;
+    }
+
+    /* Can't do anything if we don't have world */
+    if(0 == ompi_part_persist.init_world) {
+        int err = MPI_Comm_rank(MPI_COMM_WORLD, &ompi_part_persist.my_world_rank);
```

Comment:
Should change these to use the internal versions of these functions. ompi_comm vs MPI_Comm